generate.sh

```diff
@@ -65,13 +65,85 @@ import (
+# read_enum reads C enum values.
 # arg1 - enum name in C code.
 # arg2 - path to a C code.
 function read_enum {
 	grep -Pzo "(?s)enum ${1} \{.+?\};" "${2}" | grep -aP "^\t"
 }
 
+# read_define_prefix_name reads C name of the prefix for constants.
+# arg1 - name of the define with constants.
+# arg2 - enum name in C code.
+# arg3 - path to a C code.
+function read_define_prefix_name {
+  read_enum $2 $3 | grep -oP "${1}\(.+?\)" | grep -oP "\(.+?\)" | \
+    # Delete first and last symbols from the string ("(" and ")" respectively).
+    sed "s/^.//;s/.$//"
+}
+
+# read_define reads C constants from define.
+# arg1 - name of the define with constants.
+# arg2 - path to a C code.
+function read_define {
+  sed -n "/#define ${1}/,/^$/p" "${2}"
+}
+
+# read_prefix reads prefix for constants from define.
+# arg1 - name of the define, containing the prefix.
+# arg2 - path to a C code.
+function read_prefix {
+  read_define $1 $2 | grep -oP "\) .+?#" | \
+    # Delete first and last symbols from string (")" and "#" respectively).
+    # Trim spaces.
+    sed "s/^.//;s/.$//" | awk '{$1=$1};1'
+}
+
+# read_define_with_prefix reads all constants from enum (including the ones, that
+# will be generated by macros). Also reads comments.
+# arg1 - name of the define with constants.
+# arg2 - enum name in C code.
+# arg3 - path to a C code.
+function read_define_with_prefix {
+  prefix_define_name=$(read_define_prefix_name $1 $2 $3)
+  full_prefix_define=$(read_define $prefix_define_name $3)
+  prefix=$(read_prefix $prefix_define_name $3)
+  need_double=false
+  if [[ "$full_prefix_define" == *"<<"*  ]]; then
+    need_double=true
+  fi
+  enum_constants=$(read_enum $2 $3 | \
+    # Get all constants from the enum (they are not in define) with their comments.
+    grep -Pzo "[\n]?[\t ]*(\/\*)((.|\n)+?)(\*\/)\n[\t ]*[A-Z0-9_]+[\t ]+=[\t ]+[\-0-9 <]+[\-xa-f]*," | \
+    # If there are no such comments, that means grep returned error.
+    # Otherwise, we need to delete extra backslash from the end of each line.
+    sed '$!s/$/\\/' || echo "\\")
+
+  read_define $1 $3 | grep -aP "^\t" | \
+    # Mark the start of each constant (to replace extra "_(" with the prefix).
+    # Mark the end of each constants value (to delete anything after it).
+    sed -r "s/(_\()[A-Z0-9_]+[\t ,]+[\t ]+([0-9]+[\-xa-f]*)*/#&#/g" | \
+    # Mark a comma between constant and its assigned value (to replace it with the "=").
+    sed -r "s/(_\()[A-Z0-9_]+,/&#/g" | \
+    # Replace marked start with the prefix.
+    sed -r "s/#_\(/$prefix/g" | \
+    # Delete anything after a constant value (there are two cases).
+    sed -r "s/#,.*$//g" | \
+    sed -r "s/#\).*$//g" | \
+    # Replace marked comma with the "=".
+    sed -r "s/,#/ =/g" | \
+    # Delete backslash at the end of each line.
+    sed -r 's/\\//g' | \
+    # Change constants value to the power of two if needed.
+    awk '{if(f=="true") {for(i=1;i<=NF;i++) if($i+0==$i) $i=2**$i}}$0;' f=$need_double | \
```

Comment:
It would be nice just to do the same as in the original define to be less confusing:

```Go
SOME_CONST = 1 << N
```

Reply:
Updated.